File: app/src/main/java/com/emiryanvl/tapreader/ui/fragments/LibraryFragment.java
package com.emiryanvl.tapreader.ui.fragments;

import android.os.Bundle;

import androidx.recyclerview.widget.RecyclerView;

import com.emiryanvl.tapreader.App;
import com.emiryanvl.tapreader.R;
import com.emiryanvl.tapreader.databinding.FragmentLibraryBinding;
import com.emiryanvl.tapreader.ui.adapters.BookAdapter;
import com.emiryanvl.tapreader.ui.viewModels.LibraryViewModel;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import javax.inject.Inject;

public class LibraryFragment extends BaseFragment<FragmentLibraryBinding> {

    @Inject
    LibraryViewModel viewModel;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.appComponent().inject(this);
        viewModel.getBooksBySubject();
    }

    @Override
    protected Inflater<FragmentLibraryBinding> getInflater() {
        return FragmentLibraryBinding::inflate;
    }

    @Override
    protected void initializeViews() {
        recyclerView();
        floatingActionButton();
    }

    private void recyclerView() {
        RecyclerView recommendedBooksRecyclerView = binding.recommendedBooksRecyclerView;
        RecyclerView newReleasesBooksRecyclerView = binding.newReleasesBooksRecyclerView;

        viewModel.subjectBooks.observe(getViewLifecycleOwner(), bookList -> {
            BookAdapter bookAdapter = new BookAdapter(bookList);
            recommendedBooksRecyclerView.setAdapter(bookAdapter);
            newReleasesBooksRecyclerView.setAdapter(bookAdapter);
        });
    }

    private void floatingActionButton() {
        FloatingActionButton addBookfloatingActionButton = binding.addBookfloatingActionButton;
        addBookfloatingActionButton.setOnClickListener(button ->
            navController.navigate(R.id.action_libraryFragment_to_addBookFragment)
        );
    }
}

File: app/src/main/java/com/emiryanvl/tapreader/ui/fragments/LibraryFragment.kt
package com.emiryanvl.tapreader.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DiffUtil
import com.emiryanvl.tapreader.R
import com.emiryanvl.tapreader.databinding.FragmentLibraryBinding
import com.emiryanvl.tapreader.ui.adapters.BookAdapter
import com.emiryanvl.tapreader.ui.viewModels.LibraryViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class LibraryFragment : BaseFragment<FragmentLibraryBinding>() {

    private val viewModel by viewModels<LibraryViewModel>()
    override val inflate: (LayoutInflater, ViewGroup?, Boolean) -> FragmentLibraryBinding =
        FragmentLibraryBinding::inflate

    override fun initializeViews() {
        bookRecyclerView()
    }

    private fun bookRecyclerView() = with(binding) {
        val navigateOnBookTap = { bundle: Bundle ->
            navController.navigate(
                R.id.action_libraryFragment_to_bookFragment,
                bundle
            )
        }

        val bookAdapter = BookAdapter(navigateOnBookTap)

        observeFlow(viewModel.uiState) {
            val bookDiffUtilCallback = BookAdapter.Callback(bookAdapter.bookList, it.bookList)
            val productDiffResult = DiffUtil.calculateDiff(bookDiffUtilCallback)
            bookAdapter.bookList = it.bookList
            productDiffResult.dispatchUpdatesTo(bookAdapter)
        }

        recommendedBooksRecyclerView.adapter = bookAdapter
        newReleasesBooksRecyclerView.adapter = bookAdapter
    }
}